tests/sentry/receivers/outbox/test_control.py

```diff
@@ -96,8 +159,408 @@ def test_process_async_webhooks_failure(self):
+                        shard_identifier=outbox.shard_identifier,
+                        object_identifier=outbox.object_identifier,
+                    )
+                    assert len(responses.calls) == 1
+
+                    assert capture_exception.call_count == 1
+                    exception = capture_exception.call_args.args[0]
+                    assert isinstance(exception, SiloClientError)
+                    assert (
+                        exception.args[0]
+                        == f"Request attempts limit reached for: {request.method} {request.path}"
+                    )
+
+                assert mock_cache.get.call_count == 1
+                assert mock_cache.set.call_count == 0
+                assert mock_cache.delete.call_count == 1
+
+                # Assert order of cache method calls
+                expected_calls = [
+                    mock.call.cache_get(cache_key),
+                    mock.call.cache_delete(cache_key),
+                ]
+                assert parent_mock.mock_calls == expected_calls
+                return
+            else:
+                with raises(ApiError):
+                    process_async_webhooks(
+                        payload=outbox.payload,
+                        region_name=_TEST_REGION.name,
+                        shard_identifier=outbox.shard_identifier,
+                        object_identifier=outbox.object_identifier,
+                    )
+                assert len(responses.calls) == 1
+                resp = responses.calls[0].response
+                assert resp.status_code == status.HTTP_504_GATEWAY_TIMEOUT
+
+                num_of_request_attempts += 1
+
+                assert mock_cache.get.call_count == 1
+                assert mock_cache.set.call_count == 1
+                assert mock_cache.delete.call_count == 0
+
+                # Assert order of cache method calls
+                expected_calls = [
+                    mock.call.cache_get(cache_key),
+                    mock.call.cache_set(cache_key, num_of_request_attempts, timeout=CACHE_TIMEOUT),
+                ]
+                assert parent_mock.mock_calls == expected_calls
+
+    @responses.activate
+    @mock.patch("sentry.silo.client.cache")
+    def test_process_async_webhooks_retry_within_limit(self, mock_cache):
+        request, outbox, cache_key = self.generate_outbox()
+
+        responses.add(
+            request.method,
+            f"{_TEST_REGION.address}{request.path}",
+            status=status.HTTP_504_GATEWAY_TIMEOUT,
+        )
+        num_of_request_attempts = 0
+
+        while True:
+            mock_cache.reset_mock()
+            responses.calls.reset()
+
+            if num_of_request_attempts == (REQUEST_ATTEMPTS_LIMIT - 1):
+                responses.replace(
+                    request.method,
+                    f"{_TEST_REGION.address}{request.path}",
+                    status=200,
+                    json={"ok": True},
+                )
```

Comment:
You could use [responses matchers](https://pypi.org/project/responses/#creating-custom-matcher) to conditionally match the request based on the call count. Not a necessary change though.

Reply:
@markstory I think that's for matching request payloads.